src/main/java/com/google/firebase/FirebaseApp.java

```diff
@@ -388,26 +423,30 @@ FirebaseService getService(String id) {
    */
   static class TokenRefresher implements CredentialsChangedListener {
 
+    private static final int STATE_READY = 0;
+    private static final int STATE_STARTED = 1;
+    private static final int STATE_STOPPED = 2;
+
     private final FirebaseApp firebaseApp;
     private final GoogleCredentials credentials;
+    private final AtomicInteger state;
+
     private Future<Void> future;
-    private boolean closed;
 
     TokenRefresher(FirebaseApp firebaseApp) {
       this.firebaseApp = checkNotNull(firebaseApp);
       this.credentials = firebaseApp.getOptions().getCredentials();
-      this.credentials.addChangeListener(this);
+      this.state = new AtomicInteger(STATE_READY);
     }
 
     @Override
     public final synchronized void onChanged(OAuth2Credentials credentials) throws IOException {
-      if (closed) {
+      if (state.get() != STATE_STARTED) {
         return;
       }
 
       AccessToken accessToken = credentials.getAccessToken();
-      long refreshDelay = accessToken.getExpirationTime().getTime()
-          - System.currentTimeMillis() - TimeUnit.MINUTES.toMillis(5);
+      long refreshDelay = getRefreshDelay(accessToken);
       if (refreshDelay > 0) {
```

Comment:
This doesn't work for intervals <= 5 minutes.

Reply:
That is correct. However, in practice this doesn't really happen. New tokens minted by Google auth have a 1 hour TTL. If it does happen (due to some issue in the remote token server), we don't want to schedule refresh events aggressively, and cause a feedback loop. So we simply log a warning and let things run course.  This is what we have done in the past releases too.